File: bf.js
const fs = require("fs");

if (process.argv.length < 3) {
  console.log("Usage: node run.bf file.bf");
  process.exit(1);
}

const code = fs.readFileSync(process.argv[2], "utf-8").replace(/[^\[\]\+\-\.<>\,]/g, '');
let tape = new Uint8Array(30000), ptr = 0, stack = [];

for (let i = 0; i < code.length; i++) {
  let c = code[i];
  switch (c) {
    case '>': ptr++; break;
    case '<': ptr--; break;
    case '+': tape[ptr]++; break;
    case '-': tape[ptr]--; break;
    case '.': process.stdout.write(String.fromCharCode(tape[ptr])); break;
    case ',': tape[ptr] = 0; break;
    case '[':
      if (tape[ptr] === 0) {
        let loop = 1;
        while (loop > 0) {
          i++;
          if (code[i] === '[') loop++;
          if (code[i] === ']') loop--;
        }
      } else stack.push(i);
      break;
    case ']':
      if (tape[ptr] !== 0) i = stack[stack.length - 1] - 1;
      else stack.pop();
      break;
  }
}

File: bf.lua
local fname = arg[1]
if not fname then print("Usage: lua run.bf file.bf") os.exit() end

local file = io.open(fname, "r")
local code = file:read("*a"):gsub("[^%+%-%<%>%[%]%,%.]", "")
file:close()

local tape, ptr, pc, stack = {}, 1, 1, {}
for i=1, 30000 do tape[i]=0 end

while pc <= #code do
  local c = code:sub(pc, pc)
  if c == '>' then ptr = ptr + 1
  elseif c == '<' then ptr = ptr - 1
  elseif c == '+' then tape[ptr] = (tape[ptr] + 1) % 256
  elseif c == '-' then tape[ptr] = (tape[ptr] - 1) % 256
  elseif c == '.' then io.write(string.char(tape[ptr]))
  elseif c == ',' then tape[ptr] = 0
  elseif c == '[' then
    if tape[ptr] == 0 then
      local depth = 1
      while depth > 0 do
        pc = pc + 1
        local ch = code:sub(pc, pc)
        if ch == '[' then depth = depth + 1 end
        if ch == ']' then depth = depth - 1 end
      end
    else table.insert(stack, pc) end
  elseif c == ']' then
    if tape[ptr] ~= 0 then
      pc = stack[#stack]
    else table.remove(stack) end
  end
  pc = pc + 1
end

